riptide-opentelemetry/src/main/java/org/zalando/riptide/opentelemetry/OpenTelemetryPlugin.java

```diff
@@ -0,0 +1,82 @@
+package org.zalando.riptide.opentelemetry;
+
+import com.google.common.collect.Multimaps;
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import org.springframework.http.client.ClientHttpResponse;
+import org.zalando.riptide.Attribute;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestArguments;
+import org.zalando.riptide.RequestExecution;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class OpenTelemetryPlugin implements Plugin {
+
+    public static final Attribute<String> OPERATION_NAME = Attribute.generate();
+
+    private final Tracer tracer;
+    private final TextMapPropagator propagator;
+
+    private final Attribute<Span> internalSpan = Attribute.generate();
+
+    public OpenTelemetryPlugin(final Tracer tracer) {
+        this.tracer = tracer;
+        this.propagator = GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+    }
+
+    @Override
+    public RequestExecution aroundAsync(@Nonnull final RequestExecution execution) {
+        return arguments -> trace(execution, arguments);
+    }
+
+    @Override
+    public RequestExecution aroundNetwork(@Nonnull final RequestExecution execution) {
+        return arguments -> inject(execution, arguments);
+    }
+
+    private CompletableFuture<ClientHttpResponse> trace(
+            final RequestExecution execution,
+            final RequestArguments arguments) throws IOException {
+
+        final String operationName = arguments.getAttribute(OPERATION_NAME)
+                                              .orElse(arguments.getMethod().name());
+
+        final Span span = tracer.spanBuilder(operationName)
+                                .setSpanKind(SpanKind.CLIENT)
+                                .startSpan();
+
+        return execution.execute(arguments.withAttribute(internalSpan, span))
+                        .whenComplete((response, throwable) -> span.end());
+    }
+
+    private CompletableFuture<ClientHttpResponse> inject(
+            final RequestExecution execution,
+            final RequestArguments arguments) throws IOException {
+
+        @Nullable
+        final Span span = arguments.getAttribute(internalSpan)
+                                   .orElse(null);
+
+        if (span == null) {
+            return execution.execute(arguments);
+        }
+
+        try (final Scope ignored = span.makeCurrent()) {
```

Comment:
This is something we want to do right after creating the span so we'll have the contex available within the thread local scope.

Reply:
Not necessarily. Activating a span only really works in non-reactive systems. Unless you have an alternative implementation of the ScopeManager or whatever it's called in OTel now.